File: src/main/java/net/casualuhc/uhcmod/command/ReadyCommand.java
package net.casualuhc.uhcmod.command;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.context.CommandContext;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
import net.casualuhc.uhcmod.managers.GameManager;
import net.casualuhc.uhcmod.utils.uhc.Phase;
import net.casualuhc.uhcmod.managers.PlayerManager;
import net.casualuhc.uhcmod.managers.TeamManager;
import net.minecraft.scoreboard.AbstractTeam;
import net.minecraft.scoreboard.Team;
import net.minecraft.server.command.ServerCommandSource;
import net.minecraft.text.MutableText;
import net.minecraft.text.Text;
import net.minecraft.util.Formatting;

import static net.minecraft.server.command.CommandManager.literal;

public class ReadyCommand {
	private static final SimpleCommandExceptionType NOT_NOW = new SimpleCommandExceptionType(Text.translatable("uhc.lobby.ready.notNow"));
	private static final SimpleCommandExceptionType NO_TEAM = new SimpleCommandExceptionType(Text.translatable("uhc.lobby.ready.noTeam"));
	private static final SimpleCommandExceptionType ALREADY_READY = new SimpleCommandExceptionType(Text.translatable("uhc.lobby.ready.alreadyReady"));

	public static void register(CommandDispatcher<ServerCommandSource> dispatcher) {
		dispatcher.register(literal("ready")
			.then(literal("yes")
				.executes(context -> ready(context, true))
			)
			.then(literal("no")
				.executes(context -> ready(context, false))
			)
			.then(literal("reset").requires(source -> source.hasPermissionLevel(4))
				.executes(context -> {
					context.getSource().getServer().getScoreboard().getTeams().forEach(team -> TeamManager.setReady(team, false));
					return 1;
				})
			)
			.then(literal("awaiting").requires(source -> source.hasPermissionLevel(4))
				.executes(context -> {
					if (!GameManager.isPhase(Phase.READY)) {
						throw NOT_NOW.create();
					}
					ServerCommandSource source = context.getSource();
					source.sendFeedback(Text.literal("The following teams are not ready:"), true);
					for (Team team : TeamManager.getUnreadyTeams()) {
						source.sendFeedback(team.getFormattedName(), true);
					}
					return 1;
				})
			)
		);
	}

	private static int ready(CommandContext<ServerCommandSource> context, boolean isReady) throws CommandSyntaxException {
		if (!GameManager.isPhase(Phase.READY)) {
			throw NOT_NOW.create();
		}
		AbstractTeam team = context.getSource().getPlayerOrThrow().getScoreboardTeam();
		if (TeamManager.shouldIgnoreTeam(team)) {
			throw NO_TEAM.create();
		}
		if (TeamManager.isReady(team)) {
			throw ALREADY_READY.create();
		}
		TeamManager.setReady(team, isReady);
		MutableText readyText = isReady ? Text.translatable("uhc.lobby.ready.isReady", team.getName()) : Text.translatable("uhc.lobby.ready.notReady", team.getName());
		PlayerManager.messageEveryPlayer(readyText.formatted(team.getColor(), Formatting.BOLD));
		TeamManager.checkAllTeamsReady();
		return 1;
	}
}

File: src/main/kotlin/net/casualuhc/uhcmod/commands/ReadyCommand.kt
package net.casualuhc.uhcmod.commands

import com.mojang.brigadier.CommandDispatcher
import com.mojang.brigadier.context.CommandContext
import com.mojang.brigadier.exceptions.SimpleCommandExceptionType
import net.casualuhc.arcade.utils.PlayerUtils
import net.casualuhc.arcade.utils.TeamUtils.asPlayerTeam
import net.casualuhc.uhcmod.extensions.TeamFlag.Ignored
import net.casualuhc.uhcmod.extensions.TeamFlag.Ready
import net.casualuhc.uhcmod.extensions.TeamFlagsExtension.Companion.flags
import net.casualuhc.uhcmod.managers.TeamManager
import net.casualuhc.uhcmod.managers.UHCManager
import net.casualuhc.uhcmod.util.Texts
import net.minecraft.commands.CommandSourceStack
import net.minecraft.commands.Commands
import net.minecraft.network.chat.Component

object ReadyCommand: Command {
    private val NOT_NOW = SimpleCommandExceptionType(Texts.LOBBY_READY_NOT_NOW)
    private val NO_TEAM = SimpleCommandExceptionType(Texts.LOBBY_READY_NO_TEAM)
    private val ALREADY_READY = SimpleCommandExceptionType(Texts.LOBBY_READY_ALREADY)

    override fun register(dispatcher: CommandDispatcher<CommandSourceStack>) {
        dispatcher.register(
            Commands.literal("ready").then(
                Commands.literal("yes").executes {
                    this.ready(it, true)
                }
            ).then(
                Commands.literal("no").executes {
                    this.ready(it, false)
                }
            ).then(
                Commands.literal("reset").requires {
                    it.hasPermission(4)
                }.executes(this::unreadyAll)
            ).then(
                Commands.literal("awaiting").requires {
                    it.hasPermission(4)
                }.executes(this::awaiting)
            )
        )
    }

    private fun ready(context: CommandContext<CommandSourceStack>, ready: Boolean): Int {
        if (!UHCManager.isReadyPhase()) {
            throw NOT_NOW.create()
        }
        val team = context.source.playerOrException.team
        if (team === null || team.flags.has(Ignored)) {
            throw NO_TEAM.create()
        }
        if (team.flags.has(Ready)) {
            throw ALREADY_READY.create()
        }
        team.flags.set(Ready, ready)
        val generator = (if (ready) Texts.LOBBY_IS_READY else Texts.LOBBY_NOT_READY)
        val message = generator.generate(team.asPlayerTeam().formattedDisplayName)
        PlayerUtils.broadcast(message)
        TeamManager.checkAllTeamsReady()
        return 1
    }

    private fun unreadyAll(context: CommandContext<CommandSourceStack>): Int {
        context.source.server.scoreboard.playerTeams.forEach { team ->
            team.flags.set(Ready, false)
        }
        return 1
    }

    private fun awaiting(context: CommandContext<CommandSourceStack>): Int {
        if (!UHCManager.isReadyPhase()) {
            throw NOT_NOW.create()
        }
        context.source.sendSuccess(Component.literal("The following teams are not ready:"), true)
        for (team in TeamManager.getUnreadyTeams()) {
            context.source.sendSuccess(team.displayName, true)
        }
        return 1
    }
}